batchproducer/batchproducer.go

```diff
@@ -187,15 +200,21 @@ func (b *batchProducer) Add(data []byte, partitionKey string) error {
 
 // from/for interface Producer
 func (b *batchProducer) Start() error {
-	if b.isRunning() {
-		return nil
+	b.runningMu.Lock()
+	defer b.runningMu.Unlock()
```

Comment:
> This is throwing full exclusion contention around the entire method. We can reduce this to just:
>
> [code omitted]
>
> This won't make much difference in the end as spinning off a go routine and waiting for it to start won't take that long (for now, anyway.) but it feels like unnecessary contention.

This is actually OK, given the intended usage patterns of this object; users are supposed to call `Start`, wait for it to finish, and only then start calling `Add` repeatedly. There’d be no reason for them to call `Start` once its already started; this will result in an error and I don’t think it matters if it incurs slightly more contention than absolutely necessary.

Also I think the current code is less code and easier to read.

> It does look ugly. That's why atomics are a much cleaner (and lower-contention) solution to this problem.

This looks ugly to you? I find it very straightforward and easy to reason about — and I think more idiomatic than using compare and swap.